### ISSUE TITLE
Failing build on macOS

### ISSUE THREAD

**Issue by herbygillot** [ldn-softdev/jtc#8]
Seeing build failure on a macOS build environment.

Environment details:
```
macOS 10.13 (darwin/17.7.0) arch i386
Xcode 10.1
SDK 10.13
```

Command:
```
c++ jtc.cpp -o jtc -Wall -std=c++14 -Ofast
```

Build output:
```
In file included from jtc.cpp:9:
In file included from ./lib/Json.hpp:479:
./lib/Streamstr.hpp:45:32: error: use of overloaded operator '+=' is ambiguous (with operand types 'std::string' (aka 'basic_string<char, char_traits<char>, allocator<char> >') and 'void')
                             s += {&hb_[offset], cur_ - offset};
                             ~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:920:45: note: candidate function
    _LIBCPP_INLINE_VISIBILITY basic_string& operator+=(const basic_string& __str) {return append(__str);}
                                            ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/string:921:45: note: candidate function
    _LIBCPP_INLINE_VISIBILITY basic_string& operator+=(__self_view __sv)          {return append(__sv);}
                                            ^
1 error generated.
```

**Comment by ldn-softdev** [comment]
Hi,
I think it's an inefficiency (may call it a bug) on the compiler's end. Try upgrading to the latest version of xcode.

Otherwise (if compiler upgrade is not an option), try replacing that line (in `Streamstr.hpp` line 45):
```
                             s += {&hb_[offset], cur_ - offset};
```
 with the explicit constructor call:
```
                             s += std::string{&hb_[offset], cur_ - offset};
```
the former option (compiler upgrade) is preferable, b/c after fixing that line, you may face similar errors down the code.

Let me know if either of options works,
/Dmitry

**Comment by herbygillot** [comment]
Appreciated, will try this out.

This is the result of building packages of `jtc` for MacPorts, which targets numerous macOS versions and older versions of XCode, so we may have to go with your latter option.

**Comment by herbygillot** [comment]
Just an update - full tests haven't finished running, but this change looks to have resolved what we were seeing earlier.

**Comment by ldn-softdev** [comment]
good, let me know if it compiles successfully, I'll then convert all implicit list initializer constructors  throughout my entire code to explicit calls (for compatibility reason).

**Comment by herbygillot** [comment]
OK, the operator explicit cast fix works perfectly against all XCode and macOS versions tested; thank you for that.

But we do have something else breaking for XCode 8.3 and 7.3:
```
jtc.cpp:242:37: error: call to 'abs' is ambiguous
                                    abs(opt_[CHR(OPT_IND)]): 1)
                                    ^~~
/usr/include/stdlib.h:129:6: note: candidate function
int      abs(int) __pure2;
         ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/cstdlib:167:44: note: candidate function
inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
                                           ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/cstdlib:169:44: note: candidate function
inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
                                           ^
1 error generated.
```

**Comment by herbygillot** [comment]
(You can follow the status of `jtc`'s port candidate here: https://github.com/macports/macports-ports/pull/5462)

**Comment by ldn-softdev** [comment]
that one is easy to work-around, again via an explicit type cast - older compiler it seems cannot decide on the unsigned type conversion for signed overloads, so, replace the line with this:
 ```
                                    abs(static_cast<int>(opt_[CHR(OPT_IND)].hits())): 1)
```
see if it helps.

**Comment by herbygillot** [comment]
Thank you, and I appreciate your patience on this one, Dmitry.  With that change, it looks like we're passing on everything except the oldest platform in the test matrix, XCode 7.3, [as seen here](https://travis-ci.org/macports/macports-ports/builds/596111821).

Things are looking good overall, but here's what we see in 7.3:
```
In file included from jtc.cpp:9:
./lib/Json.hpp:3726:21: error: default initialization of an object of const type 'const Jnode' without a user-provided default constructor
 static const Jnode not_original;
                    ^
                                {}
./lib/Json.hpp:3751:21: error: default initialization of an object of const type 'const Jnode' without a user-provided default constructor
 static const Jnode original;
                    ^
                            {}
2 errors generated.
```

**Comment by ldn-softdev** [comment]
yes, that one is [another bug](http://open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#253) of an old compiler. add the default initializer to the statements:
```
 static const Jnode not_original{};
```
and
```
 static const Jnode original{};
```
respectively.

**Comment by ldn-softdev** [comment]
this is [another known bug](http://open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#253) of an old compiler.
Add default initializer to each lines respectively:
```
 static const Jnode not_original{};
```
and
```
 static const Jnode original{};
```

Let me know if it fixes.

**Comment by herbygillot** [comment]
Ok yes, [we are all green](https://travis-ci.org/macports/macports-ports/builds/596216773?utm_source=github_status&utm_medium=notification) . Thank you for your help here Dmitry.
We patch files with all these fixes, which we will remove if you add them for the next version of `jtc`

I'll close this issue.

**Comment by ldn-softdev** [comment]
Hi, my prior suggestion:
```
                                   abs(static_cast<int>(opt_[CHR(OPT_IND)].hits())): 1)
```
in fact  broke indention, this one would be a correct fix:
```
                                    abs(static_cast<int>(opt_[CHR(OPT_IND)])): 1)
```
sorry, for that (I ran UT only now and discovered that one goof-up)

**Comment by herbygillot** [comment]
You are correct, and that fix does indeed resolve the issue.  PR for patch fix has been opened.  In the meanwhile, one can now install `jtc` using MacPorts on macOS.  I'll open a PR updating documentation.

**Comment by ldn-softdev** [comment]
thanks a lot for including `jtc` into MacPorts! other fixes should not cause any collaterals (plus my UT did not detected any other side effects). With this, I'll close this issue.

**Comment by herbygillot** [comment]
No problem, thank you for your consistent responsiveness.

**Comment by ldn-softdev** [comment]
Hi herbygillot,

terribly sorry for bugging you, but while committing today my latest update (1.75), I have accidentally rolled back my commits to the sate before you updated the main page with MacPorts installation instructions.

May I kindly ask you to update it again? Also, let me know if the latest version (1.75) compiles for MacPorts (this is has lots of new features and updates).  

thank you,

**Comment by ldn-softdev** [comment]
Hi herbygillot,

terribly sorry for bugging you, but while committing today my latest update (1.75), I have accidentally rolled back my commits to the sate before you updated the main page with MacPorts installation instructions.

May I kindly ask you to update it again? Also, let me know if the latest version (1.75) compiles for MacPorts (this is has lots of new features and updates).

thank you,

**Comment by herbygillot** [comment]
Hello ldn, no bother at all.  I will have a PR for you soon.

I already had a Port update for `1.75a`, and this port update was without the patches we discussed in this thread as I was making the assumption that new code would have included the fixes; but obviously that was a mistaken assumption 😂: https://github.com/macports/macports-ports/pull/6058

Here are the current list of patches for jtc 1.74: https://github.com/macports/macports-ports/tree/master/textproc/jtc/files

**Comment by ldn-softdev** [comment]
My bad, indeed. Let me commit all those incompatibilities right away, I'll let you know once I'm done.

**Comment by ldn-softdev** [comment]
In fact, out of 3 patches, I missed only the last one (`patch-streamstr-hpp.diff`), now it's there. Let me know if any other issues.

**Comment by ldn-softdev** [comment]
Actually, hold on a bit. I have discovered a few minor designs overlooks in the newly introduced features, let me do a couple of quick fixes, then I'll publish the release 1.75b, then it'll be ready for MacPorts. I'll ping you when done.

**Comment by herbygillot** [comment]
No worries, take your time - I'll likely have you sorted for tomorrow

**Comment by herbygillot** [comment]
Alright, PR up: ldn-softdev/jtc/pull/12

**Comment by ldn-softdev** [comment]
Hi, I have committed my intended fixes, so the source should be all good for the MacPorts. Let me know if any problems and thanks for your MP commits.

**Comment by herbygillot** [comment]
Thank you Dmitry, branch `1.75b` is building fine without any patches.  The PR is up in MacPorts, checks look good, and we're waiting for it to be approved & merged: https://github.com/macports/macports-ports/pull/6058

**Comment by herbygillot** [comment]
OK, this was merged in, and builds fine most everywhere, but there were 2 build failures out of the full build matrix (macOS 10.6 up to 10.14):

* macOS 10.10: https://build.macports.org/builders/ports-10.10_x86_64-builder/builds/106427/steps/install-port/logs/stdio

* macOS 10.11: https://build.macports.org/builders/ports-10.11_x86_64-builder/builds/104913/steps/install-port/logs/stdio

**Comment by ldn-softdev** [comment]
Hi, let me fix those issues, most likely will do it tomorrow - I will let you know. In meanwhile Happy NY and best wishes. :)

**Comment by herbygillot** [comment]
No worries, much appreciated Dmitry.  Wishing you the same, may 2020 bring you increased success, prosperity, peace of mind and happiness.

**Comment by ldn-softdev** [comment]
Hi Herby, I have committed my last changes (v1.75c). Along with minor improvements and fixes it also should fix those compiler compatibility warnings and errors.
Let me know if any other issues.

**Comment by herbygillot** [comment]
OK excellent, thanks; PR is up (https://github.com/macports/macports-ports/pull/6102) and we'll wait to see how things look once it's merged.

**Comment by herbygillot** [comment]
Hello Dmitry; just a heads-up: the fix you proposed for 1.75c seems to be working across all compilers:

https://github.com/macports/macports-ports/commit/acd3543a970d2b4206e92c5f7394b5acc6942056  (click the green checkmark at the top-left)

We're looking good here, so I will close this issue.
